xwiki-commons-core/xwiki-commons-collection/src/main/java/org/xwiki/collection/internal/PriorityEntries.java

```diff
@@ -0,0 +1,191 @@
+     * key.
+     * 
+     * @param key the key whose associated value is to be returned
+     * @return the value to which the specified key is mapped, or {@code null} if this there is no mapping for the key
+     */
+    public E get(String key)
+    {
+        this.lock.readLock().lock();
+
+        try {
+            return this.map.get(key);
+        } finally {
+            this.lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Returns {@code true} if this map contains no key-value mappings.
+     *
+     * @return {@code true} if this map contains no key-value mappings
+     */
+    public boolean isEmpty()
+    {
+        this.lock.readLock().lock();
+
+        try {
+            return this.map.isEmpty();
+        } finally {
+            this.lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * Returns the number of key-value mappings in this map. If the map contains more than {@code Integer.MAX_VALUE}
+     * elements, returns {@code Integer.MAX_VALUE}.
+     *
+     * @return the number of key-value mappings in this map
+     */
+    public int size()
+    {
+        this.lock.readLock().lock();
+
+        try {
+            return this.map.size();
+        } finally {
+            this.lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * @return the values sorted by natural order
+     */
+    public List<E> getSorted()
+    {
+        if (this.sorted == null) {
+            this.lock.readLock().lock();
+
+            try {
+                this.sorted = this.map.values().stream().sorted().collect(Collectors.toList());
+            } finally {
+                this.lock.readLock().unlock();
+            }
+        }
+
+        return this.sorted;
+    }
+
+    /**
+     * Performs the given action for each value in this map until all entries have been processed or the action throws
+     * an exception.
+     * 
+     * @param action the action to be performed for each element
+     */
+    public void forEachEntry(Consumer<E> action)
+    {
+        this.lock.readLock().lock();
+
+        try {
+            this.map.values().forEach(action);
+        } finally {
+            this.lock.readLock().unlock();
+        }
+    }
+
+    /**
+     * @param key the key to insert
+     * @param value the value to insert
+     */
+    public void put(String key, E value)
+    {
+        this.lock.writeLock().lock();
+
+        try {
+            this.map.put(key, value);
+            this.sorted = null;
+        } finally {
+            this.lock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * @param key key whose mapping is to be removed from the map
+     * @return the previous value associated with {@code key}, or {@code null} if there was no mapping for {@code key}.
+     */
+    public E remove(String key)
+    {
+        this.lock.writeLock().lock();
+
+        try {
+            E value = this.map.remove(key);
+            if (value != null) {
```

Comment:
If this data structure doesn't support `null` values, it might be better to explicitly check this I think (I think the other methods should support `null` values).

Reply:
Actually, I've seen that other code explicitly puts `null` values, so I believe this `remove`-implementation is incorrect.

Reply:
Indeed, I don't have null values in the use cases I'm using this tool with, so I did not bother to take care of this, but would be better to be more explicit in other methods.